### PR TITLE
Add inheriting classes to `DocTools`

### LINKED ISSUE #63799: Documentation: @GDScript and @GlobalScope are at the bottom of the list
### Godot version

v4.0.alpha.custom_build [98c1fb396]

### System information

Windows 10

### Issue description

The `@GDScript` and `@GlobalScope` items are at the bottom of the list in the documentation search window. They used to be at the top.

image

### Steps to reproduce

* Press F1
* Look at the top and see that `@GDScript` and `@GlobalScope` are not there
* Scroll to the bottom to find them

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by AThousandShips**
Integrating a documentation hierarchy into `DocTools`

Part of the work for #77471 and #62524, see there for the discussions and background

Edit: I believe this is ready

* Fixes: #63799

### CODE REVIEW

**Comment by YuriSizov** [comment]
> Still working on deciding how to deal with the ordering searches, as this changes the previous order which put `Object` at the top, if maintaining that order is important I can rework it

Order should be alphabetical, so global scopes go at the top and then everything else is in their case-independent alphanumerical order (natural, perhaps?). This is something that I've been working on, actually, when I stumbled upon your work :) So if you actually manage to address this issue here, that's an amazing improvement!

**Comment by AThousandShips** [comment]
Will change the ordering! Sounds great

**Comment by AThousandShips** [comment]
We could use case sensitive natural sorting, but I think in this case this is sufficient, I'm not sure which order is preferred:
* Light3D
* LightmapProbe
* LightOccluder3D
or:
* Light3D
* LightOccluder3D
* LightmapProbe

**Comment by YuriSizov** [comment]
I'd say the first one makes more sense when you search for something by the name. Either way, we can try and see if there are complaints.

**Comment by AThousandShips** [comment]
I think this stage is ready for review properly now, and will continue to investigate the wider help indexing system as I am able

**Comment by AThousandShips** [comment]
Might solve #66407, will test

**Comment by akien-mga** [comment]
Would you have a screenshot of the result?

**Comment by AThousandShips** [comment]
Will provide, for the correct order of sorted inherited classes?

**Comment by akien-mga** [comment]
Well I'm not 100% sure what this PR changes but I'm assuming there's a before/after for the EditorHelp changes?

**Comment by AThousandShips** [comment]
This makes the hierarchy of inherited classes explicitly available in `DocTools`, which improves the sorting, and makes the accessing of the inheriting classes easier to access, for example when generating the doc view, it is also part of a larger documentation enhancement where the access to that is convenient

Will provide screenshot for the sorting

File: editor/editor_help.cpp [hunk 1]
```diff
@@ -800,34 +800,24 @@ void EditorHelp::_update_doc() {
...
+		for (RBSet<String, NaturalNoCaseComparator>::Element *itr = doc->inheriting[cd.name].front(); itr; itr = itr->next()) {
+			if (itr->prev()) {
+				class_desc->add_text(" , ");
 			}
+
+			_add_type_icon(itr->get(), theme_cache.doc_font_size, "ArrowRight");
+			class_desc->add_text(non_breaking_space); // Otherwise icon borrows hyperlink from _add_type().
+			_add_type(itr->get());
 		}
 		_pop_normal_font();
```
Comment (AThousandShips) [inline editor/editor_help.cpp:817]: This is the relevant change for that, which also avoids scanning the entire class list for all classes just to find inheriting ones

File: editor/doc_tools.cpp [hunk 2]
```diff
@@ -310,6 +310,12 @@ void DocTools::merge_from(const DocTools &p_data) {
 void DocTools::remove_from(const DocTools &p_data) {
```
Comment (YuriSizov) [inline editor/doc_tools.cpp:310]: This method doesn't seem to be used anywhere. I would remove it.

File: editor/editor_help.cpp [hunk 3]
```diff
@@ -813,34 +813,24 @@ void EditorHelp::_update_doc() {
 	}
 
 	// Descendents
```
Comment (YuriSizov) [inline editor/editor_help.cpp:815]:
```suggestion
	// Descendants
```

Unrelated, but I know you wouldn't miss an opportunity here :)

File: editor/editor_help_search.cpp [hunk 4]
```diff
@@ -331,81 +337,103 @@ bool EditorHelpSearch::Runner::_phase_match_classes_init() {
...
-		matches[class_doc.name] = match;
+		matches[class_doc->name] = match;
 	}
 
-	++iterator_doc;
-	return !iterator_doc;
+	if (iterator_doc) {
+		++iterator_doc;
+		return !iterator_doc;
+	} else {
```
Comment (YuriSizov) [inline editor/editor_help_search.cpp:421]: No need for a nested `else` if the above block returns.

**Review by YuriSizov — APPROVED** [review]
This looks great to me, love the [new ordered list](https://github.com/godotengine/godot/assets/11782833/7f77ea04-5d9e-4527-9cac-513cfb433ecb) very much! 

Some minor notes remain, and I'm also not great with data structures, so maybe @KoBeWi could take a quick look. But overall it all seems sensible.

----
PS. We should improve at some point the relevance of what is selected as the closest match, I think. Typing "path" with "Display All", for example, selects the first complete match which is a property of a navigation server struct. Very unlikely to be relevant.

godot windows editor dev x86_64_2024-01-16_14-29-02

**Comment by akien-mga** [comment]
Thanks!

**Comment by AThousandShips** [comment]
Thank you! Great to have this one done with

**Pull request merged**